# ¡Hola Ian! <a class="tocSkip"></a>

Mi nombre es David Bautista y tengo el gusto de revisar tu proyecto.

Si veo un error en la primera revisión solamente lo señalaré y dejaré que tú encuentres de qué se trata y cómo arreglarlo. Debo prepararte para que te desempeñes como especialista en Data, en un trabajo real, el responsable a cargo tuyo hará lo mismo. Si aún tienes dificultades para resolver esta tarea, te daré indicaciones más precisas en una siguiente iteración.

Te dejaré mis comentarios más abajo - **por favor, no los muevas, modifiques o borres**

Comenzaré mis comentarios con un resumen de los puntos que están bien, aquellos que debes corregir y aquellos que puedes mejorar. Luego deberás revisar todo el notebook para leer mis comentarios, los cuales estarán en rectángulos de color verde, amarillo o rojo como siguen:

<div class='alert alert-block alert-success'>
<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Muy bien! Toda la respuesta fue lograda satisfactoriamente.
</div>

<div class='alert alert-block alert-warning'>
<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Existen detalles a mejorar. Existen recomendaciones.
</div>

<div class='alert alert-block alert-danger'>

<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Se necesitan correcciones en el bloque. El trabajo no puede ser aceptado con comentarios en rojo sin solucionar.
</div>

Cualquier comentario que quieras agregar entre iteraciones de revisión lo puedes hacer de la siguiente manera:

<div class='alert alert-block alert-info'>
<b>Respuesta estudiante.</b> <a class='tocSkip'></a>
</div>

Mucho éxito en el proyecto!

<div class="alert alert-block alert-danger">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General

Hola, Ian. Espero que te encuentres muy bien.  

Te felicito por el desarrollo del proyecto hasta el momento. Ahora bien, al empezar a revisar el contenido noté un aspecto fundamental que debemos corregir para poder continuar con la revisión: en varias ocasiones vuelves a **cargar los datos desde cero**.  

Esto es un error, ya que obliga a repetir modificaciones que ya se habían aplicado previamente en cada paso. Lo ideal es **cargar los datos una sola vez al inicio del proyecto** y, a partir de allí, ir trabajando sobre el mismo objeto de datos a lo largo de todo el análisis.  

Por favor revisa este punto y realiza los ajustes necesarios para que podamos avanzar con la siguiente revisión.  

</div>


<div class="alert alert-block alert-danger">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General #2

Hola, Ian. No se logro tener en cuenta los comentarios de la revisión anterior. 
</div>
Hola David, pense que si lo habia logrado. Es que en los 3 primeros codigos que si lo deje, es porque cuando lo quito, me da error. Dice que no se ha definido pd. 

</div>
<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General #3

Hola, Ian. Gracias por tener en cuenta los diferentes comentarios. Te felicito por la culminación del proyecto. 

# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [38]:
# Importa pandas
import pandas as pd
df= ("/datasets/music_project_en.csv") 


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Perfecto, buen trabajo importando ``pandas``.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [39]:
# Lee el archivo y almacénalo en df

df = pd.read_csv("/datasets/music_project_en.csv")


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

¡Perfecto! Muy buen trabajo cargando el dataset principal para el desarrollo del proyecto.
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [40]:
# Obtén las 10 primeras filas de la tabla df


print(df.head(10))


     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Buen trabajo usando el método <code>head()</code>. Como tip, recuerda que en los notebooks de Jupyter se puede usar <code>display()</code> en lugar de <code>print()</code> para mostrar la salida de una tabla; esto genera una visualización mucho más estética. 
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [41]:
# Obtén la información general sobre nuestros datos


print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<div class="alert alert-block alert-success">
<b>Comentario de Revisor  #3</b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método ``info()``.   </div>




<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. `Identifica tú mismo el segundo problema y escríbelo aquí.`




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1.# object, lo sabemos porque en dtype dice el tipo de datos que hay.

2.# hay suficiente

3.# no estan homologados los titulos de los datos. Entre que son lower y upper y que hay espacios extra.

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [42]:
# Muestra los nombres de las columnas

print(df.columns)



Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Perfecto, buen trabajo mostrando los encabezados del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [43]:
# Bucle que itera sobre los encabezados y los pone todos en minúsculas
for col in df.columns:
    col = col.lower()

print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [44]:
# Bucle que itera sobre los encabezados y elimina los espacios


nuevos_nombres=[]

for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres

print(df.columns)


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [45]:

# Cambia el nombre de la columna "userid"
nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
print(df.columns)





Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [46]:

# Comprueba el resultado: lista de encabezados

nuevos_nombres=[]

for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)

print(df.columns)




Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Perfecto, buen trabajo modificando los encabezados del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.


### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.


In [47]:


# Calcula el número de valores ausentes

nuevos_nombres=[]
print(df.isnull().sum())
print(df.info())






user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    63736 non-null  object
 2   artist   57512 non-null  object
 3   genre    63881 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3 </b> <a class="tocSkip"></a>

Perfecto, buen trabajo identificando la cantidad de valores nulos presentes en cada columna del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [48]:


# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
nuevos_nombres=[]
print(df.isnull().sum())
print(df.info())
columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna("unknown")
    

    



user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    63736 non-null  object
 2   artist   57512 non-null  object
 3   genre    63881 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [49]:


# Cuenta los valores ausentes


print(df.isnull().sum())
print(df.info())
nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres

df.rename(columns={"userid":"user_id"}, inplace=True)
print(df.columns)
columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna("unknown")
print(df.isnull().sum())

    


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    65079 non-null  object
 2   artist   65079 non-null  object
 3   genre    65079 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None
Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')
user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Excelente, buen trabajo usando el bucle con el método <code>fillna()</code> para solucionar el problema de los valores nulos dentro del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [50]:
# Cuenta los duplicados explícitos

print(df.duplicated())
print(df.isnull().sum())
print(df.info())


0        False
1        False
2        False
3        False
4        False
         ...  
65074    False
65075    False
65076    False
65077    False
65078    False
Length: 65079, dtype: bool
user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    65079 non-null  object
 2   artist   65079 non-null  object
 3   genre    65079 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Perfecto, buen trabajo identificando la cantidad de valores duplicados explicitos dentro del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [51]:
# Elimina los duplicados explícitos

df=df.drop_duplicates()
print(df.duplicated().sum())


0


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método <code>drop_duplicates()</code> para solucionar el problema de los valores duplicados en el dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [52]:
# Comprueba de nuevo si hay duplicados

df=df.drop_duplicates()
print(df.duplicated().sum())

0


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [53]:


# Inspecciona los nombres de géneros únicos
genres_clean = df["genre"].dropna().unique()

print("Géneros únicos (sin ordenar, con NaN):")
print(df["genre"].unique())

print("\nGéneros únicos (sin ordenar, sin NaN):")
print(genres_clean)

print("\nGéneros únicos (ordenados alfabéticamente):")
print(sorted(genres_clean))






Géneros únicos (sin ordenar, con NaN):
['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetan

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.

1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:


* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.




In [54]:
# Función para reemplazar los duplicados implícitos

def replace_wrong_values(df,column,wrong_values,correct_value):
    for wrong_value in wrong_values:
        df[column]=df[column].replace(wrong_value,correct_value)
    return df

df = replace_wrong_values(df, 'genre',['hip','hop','hip-hop'],'hiphop')
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [55]:
# Elimina los duplicados implícitos

def replace_wrong_values(df,column,wrong_values,correct_value):
    for wrong_value in wrong_values:
        df[column]=df[column].replace(wrong_value,correct_value)
    return df

df = replace_wrong_values(df, 'genre',['hip','hop','hip-hop'],'hiphop')
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [56]:
# Comprueba de nuevo los duplicados implícitos
# Elimina los duplicados implícitos

def replace_wrong_values(df,column,wrong_values,correct_value):
    for wrong_value in wrong_values:
        df[column]=df[column].replace(wrong_value,correct_value)
    return df

df = replace_wrong_values(df, 'genre',['hip','hop','hip-hop'],'hiphop')
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Excelente, buen trabajo estructurando y aplicando la función para solucionar el problema de los valores duplicados implícitos en la columna con información de género.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [57]:
# Cuenta las canciones reproducidas en cada ciudad

nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
print(df.groupby("city")["track"].count())





city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.


Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse
Springfield escucha mas musica que Shelbyville


Escribe tus observaciones aquí. springfield escucho mas de el doble de canciones que shelbyville


Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.




In [58]:
# Calcula las canciones reproducidas en cada uno de los dos días


nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
print(df.groupby("day")["track"].count())
print(df['day'].unique())
df_filtered=df[(df['city'].isin(['Springfield','Shelbyville']))&(df['day'].isin(['Monday','Friday']))]
print(df_filtered)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64
['Wednesday' 'Friday' 'Monday']
        user_id                              track                   artist  \
1      55204538        Delayed Because of Accident         Andreas Rönnberg   
3      A3DD03C9              Dragons in the Sunset               Fire + Ice   
4      E2DC1FAE                        Soul People               Space Echo   
5      842029A1                             Chains                 Obladaet   
8      8FA1D3BE                           L’estate              Julia Dalia   
...         ...                                ...                      ...   
65073  83A474E7      I Worship Only What You Bleed  The Black Dahlia Murder   
65075  D08D4A55  Maybe One Day (feat. Black Spade)              Blu & Exile   
65076  C5E3A0D5                          Jalopiina                  unknown   
65077  321D0506                      Freight Train            Chas McDevitt   
65078  3A64E

<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. 

Escribe tus observaciones aquí.
Describe brevemente qué observaste al comparar los lunes y viernes.
en viernes se escucharon 662 canciones mas que en lunes
¿Hubo un día con más actividad? si, el viernes
¿Cambia algo si analizas cada ciudad por separado?
  genre         city      time     day  
1              rock  Springfield  14:07:09  Friday  
3              folk  Shelbyville  08:37:09  Monday  
4             dance  Springfield  08:34:34  Monday  
5            rusrap  Shelbyville  13:09:41  Friday  
8            ruspop  Springfield  09:17:40  Friday  
...             ...          ...       ...     ...  
65073  extrememetal  Springfield  21:07:12  Monday  
65075           hip  Shelbyville  10:00:00  Monday  
65076    industrial  Springfield  20:09:26  Friday  
65077          rock  Springfield  21:43:59  Friday  
65078       country  Springfield  21:59:46  Friday  

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [59]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
def number_tracks(day,city):
    day_filter=df[df["day"]==day]
    city_day_filter=day_filter[day_filter['city']==city]
    track_count=city_day_filter['user_id'].count() 
    return track_count


    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=

    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()

    # Devuelve el número de valores de la columna 'user_id'

<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [60]:


nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
def number_tracks(day,city):
    day_filter=df[df["day"]==day]
    city_day_filter=day_filter[day_filter['city']==city]
    track_count=city_day_filter['user_id'].count() 
    return track_count

print("Springfield, Monday:", number_tracks('Monday', 'Springfield'))


Springfield, Monday: 15740


In [61]:
# El número de canciones reproducidas en Shelbyville el lunes


nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
def number_tracks(day,city):
    day_filter=df[df["day"]==day]
    city_day_filter=day_filter[day_filter['city']==city]
    track_count=city_day_filter['user_id'].count() 
    return track_count
print("Shelbyville, Monday:", number_tracks('Monday', 'Shelbyville'))



Shelbyville, Monday: 5614


In [62]:
# El número de canciones reproducidas en Springfield el viernes

nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
def number_tracks(day,city):
    day_filter=df[df["day"]==day]
    city_day_filter=day_filter[day_filter['city']==city]
    track_count=city_day_filter['user_id'].count() 
    return track_count
print("Springfield, Friday:", number_tracks('Friday', 'Springfield'))

Springfield, Friday: 15945


In [63]:
# El número de canciones reproducidas en Shelbyville el viernes

nuevos_nombres=[]
for col in df.columns:
    nuevo_nombre = col.lower().strip().replace(" ", "_")
    nuevos_nombres.append(nuevo_nombre)   
df.columns=nuevos_nombres
df.rename(columns={"userid":"user_id"}, inplace=True)
def number_tracks(day,city):
    day_filter=df[df["day"]==day]
    city_day_filter=day_filter[day_filter['city']==city]
    track_count=city_day_filter['user_id'].count() 
    return track_count
print("Shelbyville, Friday:", number_tracks('Friday', 'Shelbyville'))

Shelbyville, Friday: 5895


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Buen trabajo estructurando y aplicando la función.  Simplemente ten en cuenta que podrías evitar duplicar código para hacerla más eficiente y legible. 
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


Detalla aquí tus conclusiones.

Hay una clara tendencia a escuchar mas musica en viernes que en lunes. Si bien, la variacion no es masiva, es notoria.Tambin se puede apreciar que en Springfield se escuch casi 3 veces mas musica que en Shebyville, pero que la cantidad de variacion de audioescuchas en los viernes es mayor en Shelbyvlle que en Springfield (en SVille son 277 mas que en lunes y en Springfield son 175 mas que en lunes).
Los principales problemas fueron lo datos ausentes y similares. Eso se resuelve con limpieza de datos.
Respondiendo a la pregunta "*¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*" Si, como decia anteriormente, hay variacion entre dias y ciudades. En springfield se escuchan 1671 canciones en lunes y en shelbyville 5982. En viernes en springfield se escuchan 16890 canciones y en shelbyville 6259.

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #3</b> <a class="tocSkip"></a>

Buen trabajo. </div>
